ibis/backends/pandas/execution/arrays.py

```diff
@@ -44,28 +45,62 @@ def execute_array_index(op, data, index, **kwargs):
     )
 
 
-@execute_node.register(ops.ArrayIndex, list, int)
+@execute_node.register(ops.ArrayIndex, np.ndarray, int)
 def execute_array_index_scalar(op, data, index, **kwargs):
     try:
         return data[index]
     except IndexError:
         return None
 
 
-@execute_node.register(ops.ArrayConcat, pd.Series, (pd.Series, list))
-@execute_node.register(ops.ArrayConcat, list, pd.Series)
-@execute_node.register(ops.ArrayConcat, list, list)
-def execute_array_concat(op, left, right, **kwargs):
-    return left + right
+def _concat_iterables_to_series(
+    iter1: Iterable[Any], iter2: Iterable[Any],
+) -> pd.Series:
+    """Concatenate two iterables elementwise ("horizontally") to create a
+    Series. The two iterables are assumed to have the same length.
 
+    Used for ArrayConcat implementation.
+    """
+    assert len(iter1) == len(iter2)
+    result = pd.Series(map(lambda x, y: np.concatenate([x, y]), iter1, iter2))
+    return result
 
-@execute_node.register(ops.ArrayRepeat, pd.Series, pd.Series)
-@execute_node.register(ops.ArrayRepeat, int, (pd.Series, list))
-@execute_node.register(ops.ArrayRepeat, (pd.Series, list), int)
-def execute_array_repeat(op, left, right, **kwargs):
-    return left * right
+
+@execute_node.register(ops.ArrayConcat, pd.Series, pd.Series)
+def execute_array_concat_series(op, left, right, **kwargs):
+    return _concat_iterables_to_series(left, right)
+
+
+@execute_node.register(ops.ArrayConcat, pd.Series, np.ndarray)
+@execute_node.register(ops.ArrayConcat, np.ndarray, pd.Series)
+def execute_array_concat_mixed(op, left, right, **kwargs):
+    if isinstance(left, np.ndarray):
+        # Broadcast `left` to the length of `right`
+        left = np.tile(left, (len(right), 1))
+    elif isinstance(right, np.ndarray):
+        # Broadcast `right` to the length of `left`
+        right = np.tile(right, (len(left), 1))
+    return _concat_iterables_to_series(left, right)
+
+
+@execute_node.register(ops.ArrayConcat, np.ndarray, np.ndarray)
+def execute_array_concat_scalar(op, left, right, **kwargs):
+    return np.concatenate([left, right])
+
+
+@execute_node.register(ops.ArrayRepeat, pd.Series, int)
+def execute_array_repeat(op, data, n, **kwargs):
+    # Negative n will be treated as 0 (repeat will produce empty array)
```

Comment:
Hmm... why do we allow negative n at all?

```
np.repeat(v, -1)
```
will raise an error

Reply:
Hmm that's a good point. I can raise an error instead

Reply:
I'm actually now thinking of keeping this as it is. It seems like this was allowed because `[1, 2, 3] * -3` produces  `[]`, and also the implementations of some of the other backends (e.g. PySpark, Postgres) naturally allow negative `n`.

(We could add extra validation logic somewhere to not allow negative `n` across the board, but we wouldn't be able to add the validation at the API/operation level—we need to know the materialized value of `n` to check if its negative, so the validation would have to be in each individual backend's execution code. All this to say: I think disallowing negative `n` is more complex than just making the Pandas backend support negative `n`)

Please let me know what you think

Reply:
Sounds good to me.